### PR TITLE
[MU4] Implemented missing file operations

### PR DESCRIPTION
**Pull request by Eism**

### CODE REVIEW

File: src/framework/uicomponents/uicomponentstypes.h [hunk 1]
```diff
@@ -57,7 +57,8 @@ struct MenuItem : public actions::ActionItem
             { "icon", static_cast<int>(iconCode) },
             { "enabled", enabled },
             { "checked", checked },
-            { "subitems", _subitems }
+            { "subitems", _subitems },
+            { "data", data }
```
Comment (RomanPudashkin) [inline src/framework/uicomponents/uicomponentstypes.h:61]: I think, better to rename data to params or args

File: src/notation/imasternotation.h [hunk 2]
```diff
@@ -41,7 +41,8 @@ class IMasterNotation
     virtual Ret createNew(const ScoreCreateOptions& scoreInfo) = 0;
     virtual RetVal<bool> created() const = 0;
 
-    virtual Ret save(const io::path& path = io::path()) = 0;
+    virtual Ret save(const io::path& path = io::path(), SaveMode saveMode = SaveMode::Unknown) = 0;
+    virtual Ret saveSelection(const io::path& path = io::path()) = 0;
```
Comment (RomanPudashkin) [inline src/notation/imasternotation.h:45]: I think, better to implement this using one method. We can create a special enum to separate two different modes

File: src/userscores/internal/filescorecontroller.cpp [hunk 3]
```diff
@@ -121,12 +127,47 @@ void FileScoreController::saveScore()
...
-    io::path selectedFilePath = selectScoreSavingFile(defaultFilePath);
+    io::path selectedFilePath = selectScoreSavingFile(defaultFilePath, qtrc("userscores", "Save Score"));
+    if (selectedFilePath.empty()) {
+        return;
+    }
+
+    doSaveScore(selectedFilePath, SaveMode::SaveAs);
+}
+
+void FileScoreController::saveScoreACopy()
```
Comment (RomanPudashkin) [inline src/userscores/internal/filescorecontroller.cpp:138]: Better to rename to saveScoreCopy

File: src/userscores/internal/userscoresservice.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    return result;
+}
+
+std::vector<Meta> UserScoresService::parseRecentList(const QStringList& recentScoresPathList) const
+{
+    std::vector<Meta> result;
+    for (const QString& path : recentScoresPathList) {
+        RetVal<Meta> meta = msczMetaReader()->readMeta(path);
+        if (!meta.ret) {
+            LOGW() << "Score reader error" << path;
```
Comment (RomanPudashkin) [inline src/userscores/internal/userscoresservice.cpp:50]: Better to use LOGE in this case

File: src/appshell/view/appmenumodel.cpp [hunk 5]
```diff
@@ -42,23 +42,18 @@ void AppMenuModel::load()
...
-    });
+    setupConnections();
 
     emit itemsChanged();
 }
 
-void AppMenuModel::handleAction(const QString& actionCode)
+void AppMenuModel::handleAction(const QString& actionCode, const QVariant& actionData)
 {
+    if (actionCode == "file-open") {
```
Comment (RomanPudashkin) [inline src/appshell/view/appmenumodel.cpp:52]: I don't like the idea to write a logic that depends on concrete codes here. This approach has poor scalability (you should add else-if for all actions that have arguments). Better write an abstract code for handling actions:

args = action.args

if (action.args.isEmpty) {
dispatch(actionCode)
} else {
dispatch(actionCode, action.args)
}

Comment (igorkorsukov) [reply src/appshell/view/appmenumodel.cpp:52]: or just `dispatch(actionCode, action.args)`

Comment (Eism) [reply src/appshell/view/appmenumodel.cpp:52]: args is QVariant, not ActionData
need to restore the types of arguments, and this is possible only by code

File: src/framework/uicomponents/uicomponentstypes.h (hunk 1 above)
Comment (Eism) [reply src/framework/uicomponents/uicomponentstypes.h]: thanks, renamed to args

File: src/notation/imasternotation.h (hunk 2 above)
Comment (Eism) [reply src/notation/imasternotation.h]: thanks, fixed

File: src/userscores/internal/filescorecontroller.cpp (hunk 3 above)
Comment (Eism) [reply src/userscores/internal/filescorecontroller.cpp]: thanks, renamed

File: src/userscores/internal/userscoresservice.cpp (hunk 4 above)
Comment (Eism) [reply src/userscores/internal/userscoresservice.cpp]: fixed

File: src/appshell/view/appmenumodel.cpp (hunk 5 above)
Comment (Eism) [reply src/appshell/view/appmenumodel.cpp]: thanks 
I am not passing arguments to qml now. The index of the action in the group comes from qml, by which I find the action locally

File: src/appshell/view/appmenumodel.cpp [hunk 6]
```diff
@@ -42,24 +43,17 @@ void AppMenuModel::load()
...
+    setupConnections();
 
     emit itemsChanged();
 }
 
-void AppMenuModel::handleAction(const QString& actionCode)
+void AppMenuModel::handleAction(const QString& actionCode, int actionIndex)
 {
-    actionsDispatcher()->dispatch(codeFromQString(actionCode));
+    ActionCode _actionCode = codeFromQString(actionCode);
```
Comment (igorkorsukov) [inline src/appshell/view/appmenumodel.cpp:53]: I believe, better in such a case  use `val_` for args

**Review by igorkorsukov — APPROVED** [review]

File: src/appshell/view/appmenumodel.cpp (hunk 6 above)
Comment (Eism) [reply src/appshell/view/appmenumodel.cpp]: renamed

**Review by RomanPudashkin — APPROVED** [review]

**Pull request merged**